Stratis.Bitcoin.Features.Wallet/WalletManager.cs

```diff
@@ -384,10 +384,9 @@ public string GetExtPubKey(WalletAccountReference accountReference)
                 HdAccount account = wallet.GetAccountByCoinType(accountReference.AccountName, this.coinType);
                 res = account.ExtendedPubKey;
 
+                this.logger.LogTrace("(-):'{0}'", res);
```

Comment:
I suggest putting the trace out of the lock.
Inside a lock we have to do the minimal possible operations that only effect the component the lock is protecting

Reply:
You are reading the resource with LogTrace.

Reply:
The `resource` you mean the parameter `res`?
That's fine it can be out of the lock, you are not manipulating it..

Reply:
> That's fine it can be out of the lock, you are not manipulating it.  

I'm reading it. Consider this: 1. lock ends, 2. other thread writes it, 3. count() calculated (wrongly!)

Reply:
I don't really mind having a log inside lock. Only look at performance when logging is at default configuration (i.e. trace logs are disabled). Then the performance hit is absolutely negligible.

Reply:
>I'm reading it. Consider this: 1. lock ends, 2. other thread writes it, 3. count() calculated (wrongly!)

If you want to trace such a count then do the count inside the lock and trace it outside, but its negligible, besides if count can be different outside of the lock then you have synchronization issues regardless of the trace, 

@Aprogiena I consider it bad practice to trace inside a lock  (unless absolutely required lets avoid it)

Reply:
I consider it bad practice to access something access something outside the lock when the lock is meant to access that something.  

> I consider it bad practice to trace inside a lock  

Indeed.  

> If you want to trace such a count then do the count inside the lock and trace it outside,  

I'd do this. @Aprogiena What do you think?

Reply:
there are different situations

1) if you put log into the critical section, the value logged will guarantee the correct state of the object and if you put it outside, it won't (i.e. it might be different) - in most cases this is not a problem to put the log outside as the logs are mostly informative, but rarely it can happen that you have a bug and you would really appreciate to have all information in logs correct, otherwise you can waste quite a lot of time wondering what just happened

the con of having the log inside the critical section is having the CS is little wider. however, in case of trace log, it is only wider of one call and one if, that's practically absolute zero performance hit. imho this performance hit is not enough reason to put the log outside

i would certainly **not** call this a bad practice

2) if you put log into the critical section, nothing changes (because the log logs something that is orthogonal to the CS), in this case naturally you would put the log outside **as a good practice** - there is no benefit in this case having the log inside CS, therefore it should not be there because CSs should be as short as possible

-----------

so which one is it here, imho this one is obviously 2) because it could be written as 

```
           string res = null;
           lock (this.lockObject)
            {
                // Get the account.
                HdAccount account = wallet.GetAccountByCoinType(accountReference.AccountName, this.coinType);
                res = account.ExtendedPubKey;
            }
            this.logger.LogTrace("(-):'{0}'", res);
            return res;
```

the `res` value can't change outside the lock because it is local variable, therefore there is no benefit to have the log inside CS, therefore it is 2) and the original suggestion of @dangershony is correct

Reply:
in this particular case, i'm not sure if even this:
```
            lock (this.lockObject)
            {
                // Get the account.
                HdAccount account = wallet.GetAccountByCoinType(accountReference.AccountName, this.coinType);
            }

            string res = account.ExtendedPubKey;
            this.logger.LogTrace("(-):'{0}'", res);
            return res;
```

would not work. the lock is said to be needed for collections over wallet, so it should have no effect on account and its ExtendedPubKey. but i'm not familiar with this code, this is just how i read the contract of lockObject

little offtopic here, but i would definitely add a comment to GetAccountByCoinType that its caller is responsible to hold this lock and i would change its name to GetAccountByCoinTypeLocked

Reply:
@dangershony Make your decision and let's wrap this small issue up.

Reply:
Yes @Aprogiena is right on the last case `ExtendedPubKey` does not need to be protected by the lock (only the lookup for account) so we just need to lock the the method (and enumeration).